### PR TITLE
ATB-1068 Added TxMA ingress schema validation

### PR DESCRIPTION
**Pull request by amd-gds**
## Proposed changes
Added json schema validation for incoming events

### Issue tracking

- [ATB-1068](https://govukverify.atlassian.net/browse/ATB-1068)

## Testing
[Testing-schma-validation.docx](https://github.com/govuk-one-login/account-interventions-service/files/13239429/Testing-schma-validation.docx)

## Checklists
- [x] Did not commit any not-required changes to the `src/infra/**/samconfig.toml`
- [x] Tested changes and included test evidence in the PR, if appropriate
- [x] Included all required tags and other properties for any new resources in the SAM template
- [x] Ensured that any new resources in the SAM Template follow appropriate naming conventions
- [x] Ensured that naming of new resources is compatible with deploying multiple stacks with custom stack names during development
- [x] Ensured that no log lines include PII or other sensitive data
- [x] Implemented unit testing for any new logic implemented, if appropriate
- [x] Ensured that all commits in this PR are signed
- [x] Ensured appropriate code coverage is maintained by unit tests

[ATB-1068]: https://govukverify.atlassian.net/browse/ATB-1068?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/data-types/schemas.ts [hunk 1]
```diff
@@ -0,0 +1,102 @@
+export const TxMAIngress = {
+  definitions: {},
+  $schema: 'http://json-schema.org/draft-07/schema#',
+  $id: 'https://example.com/object1689242392.json',
+  title: 'Encrypt Event Validator schema',
```
Comment (TaherKapasi) [inline src/data-types/schemas.ts:5]: please update to not reflect VC storage

File: src/data-types/schemas.ts [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    event: {
+      $id: '#root/event',
+      title: 'Event',
+      type: 'object',
+      required: ['timestamp', 'event_name', 'user'],
+      properties: {
+        timestamp: {
+          $id: '#root/event/timestamp',
+          title: 'Timestamp',
+          type: 'number',
```
Comment (TaherKapasi) [inline src/data-types/schemas.ts:19]: is there any extended type for timestamps?

File: src/data-types/schemas.ts (hunk 1 above)
Comment (TaherKapasi) [reply src/data-types/schemas.ts:5]: fraud risk engine intervention event

File: src/data-types/schemas.ts [hunk 3]
```diff
@@ -0,0 +1,102 @@
+export const TxMAIngress = {
```
Comment (TaherKapasi) [inline src/data-types/schemas.ts:1]: could we include the manual fraud interventions and the user led actions txma events as well - thanks

**Review by TaherKapasi — CHANGES_REQUESTED** [review]

File: src/data-types/schemas.ts [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+      $id: '#root/event',
+      title: 'Event',
+      type: 'object',
+      required: ['timestamp', 'event_name', 'user'],
+      properties: {
+        timestamp: {
+          $id: '#root/event/timestamp',
+          title: 'Timestamp',
+          type: 'number',
+          examples: [1234],
```
Comment (liamquayle) [inline src/data-types/schemas.ts:20]: small question, we removed the examples field from the schema in vc storage, just to help improve compile time, should we do the same here as a slight optimisation?

File: src/services/validate-event.ts [hunk 5]
```diff
@@ -3,37 +3,31 @@ import logger from '../commons/logger';
...
 export function validateEvent(interventionRequest: TxMAEvent): void {
-  if (
-    !interventionRequest.timestamp ||
-    Number.isNaN(interventionRequest.timestamp) ||
-    !interventionRequest.user?.user_id ||
-    !interventionRequest.event_name
-  ) {
-    logger.debug('event has failed initial validation');
+  if (!validateInterventionDataInput({ event: interventionRequest })) {
+    logger.debug('event has failed schema validation');
```
Comment (jessica-li-01) [inline src/services/validate-event.ts:18]: Can we change to capital letter and full stop 'Event has failed the schema validation.'?

File: src/services/test/validate-event.test.ts [hunk 6]
```diff
@@ -66,7 +65,29 @@ describe('event-validation', () => {
...
-    expect(logger.debug).toHaveBeenCalledWith('event has failed initial validation');
+    expect(logger.debug).toHaveBeenCalledWith('event has failed schema validation');
+    expect(logAndPublishMetric).toHaveBeenCalledWith('INVALID_EVENT_RECEIVED');
+  });
+
+  it('should return error as intervention code is NAN', () => {
+    const TxMAEvent: TxMAEvent = {
+      timestamp: 87_298_174,
+      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+      // @ts-ignore
```
Comment (chiarapascucci-gds) [inline src/services/test/validate-event.test.ts:76]: why is ts-ignore needed here?

Comment (amd-gds) [reply src/services/test/validate-event.test.ts]: good spot. it's not!

**Review by TaherKapasi — APPROVED** [review]

File: src/data-types/schemas.ts [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+          $id: '#root/event/event_name',
+          title: 'Event_Name',
+          type: 'string',
+        },
+        component_id: {
+          $id: '#root/event/component_id',
+          title: 'Component_id',
+          type: 'string',
+        },
+        extension: {
```
Comment (TaherKapasi) [inline src/data-types/schemas.ts:36]: extensions: with an 's' at the end

File: src/data-types/schemas.ts [hunk 8]
```diff
@@ -0,0 +1,147 @@
+export const TxMAIngress = {
+  definitions: {},
+  $schema: 'http://json-schema.org/draft-07/schema#',
+  $id: 'https://example.com/object1689242392.json',
```
Comment (TaherKapasi) [inline src/data-types/schemas.ts:4]: give it a better id - use the git url of the schema

File: src/data-types/schemas.ts [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+          type: 'string',
+        },
+        component_id: {
+          $id: '#root/event/component_id',
+          title: 'Component_id',
+          type: 'string',
+        },
+        extension: {
+          $id: '#root/event/extension',
+          title: 'Extension',
```
Comment (TaherKapasi) [inline src/data-types/schemas.ts:38]: Extensions

File: src/data-types/schemas.ts [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+          required: ['user_id'],
+          properties: {
+            user_id: {
+              $id: '#root/event/user_id',
+              title: 'User_Id',
+              type: 'string',
+              pattern: '^[^,\\s]+$',
+              minLength: 3,
+            },
+            email: {
```
Comment (TaherKapasi) [inline src/data-types/schemas.ts:112]: probably should ignore these attributes below as we don't need them and add `additionalProperties: true` just so we don't have to list ones we don't need.

**Review by TaherKapasi — CHANGES_REQUESTED** [review]

File: src/services/test/validate-event.test.ts [hunk 11]
```diff
@@ -11,7 +11,7 @@ describe('event-validation', () => {
     jest.clearAllMocks();
   });
 
-  it('should return true as event is valid', () => {
+  it('should no return anything as event is valid', () => {
```
Comment (jessica-li-01) [inline src/services/test/validate-event.test.ts:14]: *not

File: src/services/test/validate-event.test.ts [hunk 12]
```diff
@@ -29,7 +29,7 @@ describe('event-validation', () => {
     expect(validateInterventionEvent(TxMAEvent)).toBeUndefined();
   });
 
-  it('should return false as intervention is invalid', () => {
+  it('should return error as intervention is invalid', () => {
```
Comment (jessica-li-01) [inline src/services/test/validate-event.test.ts:32]: *an error

File: src/services/test/validate-event.test.ts [hunk 13]
```diff
@@ -44,13 +44,12 @@ describe('event-validation', () => {
...
-    expect(validateEvent(TxMAEvent)).toBeUndefined();
-    expect(() => validateInterventionEvent(TxMAEvent)).toThrow(new ValidationError('Invalid intervention event.'));
-    expect(logger.debug).toHaveBeenCalledWith('Invalid intervention request.');
+    expect(() => validateEvent(TxMAEvent)).toThrow(new ValidationError('Invalid intervention event.'));
+    expect(logger.debug).toHaveBeenCalledWith('event has failed schema validation');
     expect(logAndPublishMetric).toHaveBeenCalledWith('INVALID_EVENT_RECEIVED');
   });
 
-  it('should return false as event is invalid', () => {
+  it('should return error as event is invalid', () => {
```
Comment (jessica-li-01) [inline src/services/test/validate-event.test.ts:52]: *an error

File: src/services/test/validate-event.test.ts [hunk 14]
```diff
@@ -66,7 +65,27 @@ describe('event-validation', () => {
     };
     expect(() => validateEvent(TxMAEvent)).toThrow(new ValidationError('Invalid intervention event.'));
     expect(validateInterventionEvent(TxMAEvent)).toBeUndefined();
-    expect(logger.debug).toHaveBeenCalledWith('event has failed initial validation');
+    expect(logger.debug).toHaveBeenCalledWith('event has failed schema validation');
```
Comment (jessica-li-01) [inline src/services/test/validate-event.test.ts:68]: Can we have a capital letter *Event pls?:)

**Review by TaherKapasi — APPROVED** [review]

**Review by chiarapascucci-gds — APPROVED** [review]

**Review by jessica-li-01 — APPROVED** [review]

**Pull request merged**